server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -242,59 +262,93 @@ public Object persist(final Committer committer) throws InterruptedException
   }
 
   /**
-   * Publish all data indexed through this driver so far, and waits for it to be handed off. Blocks until complete.
-   * Retries forever on transient failures, but may exit early on permanent failures.
-   *
-   * Should be called after all data has been added and persisted through {@link #add(InputRow, String, Supplier)} and
-   * {@link #persist(Committer)}.
+   * Register the segments in the given {@link SegmentsAndMetadata} to be handed off and execute a background task which
+   * waits until the hand off completes.
    *
-   * @param publisher publisher to use for this set of segments
-   * @param committer committer representing all data that has been added so far
-   *
-   * @return segments and metadata published if successful, or null if segments could not be handed off due to
-   * transaction failure with commit metadata.
+   * @param segmentsAndMetadata
+   * @return
+   * @throws InterruptedException
    */
-  public SegmentsAndMetadata finish(
-      final TransactionalSegmentPublisher publisher,
-      final Committer committer
-  ) throws InterruptedException
+  public ListenableFuture<SegmentsAndMetadata> registerHandoff(SegmentsAndMetadata segmentsAndMetadata)
```

Comment:
I am not sure about this method. I think we need to do two things 
1. Register segment handoff callback with `SegmentHandoffNotifier` so that when successful hand off happens appenderator can drop the handed off segments.
2. Wait for all the segments in the appenderator to be handed off. That can be done by `waiting` on `handoffMonitor` which will get notified every time a callback is done from the `SegmentHandoffNotifier`. And the check will done in a loop which checks whether there are still any segments present in the appenderator.
There is already code that does both things, it is just that having two methods `registerHandoff ` and `waitForHandOff` would be useful. I agree with @gianm we do not need any executor in both methods.